### PR TITLE
Ut 190/userdetails

### PR DESCRIPTION
**Pull request by idax6797**
Branch with unit test and the 'old' UI for the editable table, see link. 

https://xd.adobe.com/view/822982cf-0e7a-4574-8e7d-34d0ca7b09d6-d77a/comment/083e220e-901b-4c6e-b733-3a79d14209e7/

### CODE REVIEW

**Comment by andr9528** [comment]
How many of the files under `frontend/coverage` is autogenerated?
It would appear as thought most, if not all, of it is autogenereated, and thus could/would be generated again, should it not be there.
 
Because of this, if it is autogenerated, i'd very much like for those files to be excluded from Git, via the `.gitignore`.
Would also remove about 1/3 of of the changes in this PR.

**Comment by KvelaGorrrrnio** [comment]
> How many of the files under `frontend/coverage` is autogenerated? It would appear as thought most, if not all, of it is autogenereated, and thus could/would be generated again, should it not be there.
> 
> Because of this, if it is autogenerated, i'd very much like for those files to be excluded from Git, via the `.gitignore`. Would also remove about 1/3 of of the changes in this PR.

The coverage files are not supposed to be versioned controlled. So this should be removed from the git tree and git ignored.

File: frontend/src/admin/__tests__/dropdown-menu.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    expect(wrapper.container).toBeInTheDocument()
+  })
+
+  it('should show/hide dropdown when button is clicked', async() => {
+    const button = wrapper.getByRole('button')
+    act(() => {
+      button.click()
+    })
+    await waitFor(() => {
+      expect(wrapper.container.querySelector('.c-dropdown-menu--active')).toBeInTheDocument()
```
Comment (LindvedKrvang) [inline frontend/src/admin/__tests__/dropdown-menu.spec.tsx:24]: Maybe also check that `.c-dropdown-menu--active` is not in the document to verify that the dropdown was removed after the second click?

File: frontend/src/admin/__tests__/user-information.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      name: 'sofie',
+      accessRights: [{
+        machineId: 'test',
+        label: 'test',
+        path: '/test',
+        anonymous: true
+      }]
+    }]}
+  it('renders without crashing', () => {
+    render(<UserInformation user={user} />)
```
Comment (LindvedKrvang) [inline frontend/src/admin/__tests__/user-information.spec.tsx:31]: In the `render without crashing` test for DropdownMenu you check `.toBeInTheDocument()`, wouldn't you also need this here?
If I understand `render` correct, then there is no assertion happening in your test and this test will never fail?

File: frontend/src/admin/__tests__/user-information.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  })
+  it('displays the correct user information', () => {
```
Comment (LindvedKrvang) [inline frontend/src/admin/__tests__/user-information.spec.tsx:35]: Minor thing: Would be nice if you left a single empty line between your tests :)

File: frontend/src/admin/__tests__/user-list-item.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      accessRights: [{
+        machineId: 'test',
+        label: 'test',
+        path: '/test',
+        anonymous: true
+      }]
+    }]}
+  it('renders without crashing', () => {
+    const onClick = jest.fn()
+    render(<UserListItem user={user} onClick={onClick} isSelected={false} />)
```
Comment (LindvedKrvang) [inline frontend/src/admin/__tests__/user-list-item.spec.tsx:23]: Again, unless I misunderstand `render` (which might be the case) then there is no assertion here. Have you seen this test fail? (If you haven't seen your tests fail, then you don't know if they work/test what you actually expect them to test)

Sidenote: empty lines between your tests please :)

File: frontend/src/admin/__tests__/user-information.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import { render, screen } from '@testing-library/react'
+import UserInformation from '../components/user-information/user-information'
+import { describe, it } from '@jest/globals'
+import '@testing-library/jest-dom'
+
+describe('UserInformation', () => {
+  const user = {
```
Comment (LindvedKrvang) [inline frontend/src/admin/__tests__/user-information.spec.tsx]: Food for thought:

Right now you are creating a single instance of `user` and parsing it to all your tests. This might lead to some nasty, potentially random, "bugs". 
Given the example below:
```
describe('some tests', () => {
  const user = { name: 'UserName' }

  it('A - changes the user name', () => {
    const result = changeUserName(user)
    expect(result.name).isEqualTo('NewUserName')
  }

  it('B - does not change the user name', () => {
    const result = doSomethingThatWontChangeTheUserName(user)
    expect(result.name).isEqualTo('UserName')
  }
})
```
In the above code; if test A runs before test B, then test B will fail since both tests are using the same instance of the User and test A changes the data of the User.
If test B runs before test A then everyhing is fine, since test A hasn't had a chance to change the data yet.
Since the tests should hopefully be run asynchronisly (they most likely aren't when you are doing component testing like you are here) then we would get randomly failing tests.

The fix would be to instance the User object for each of your tests, like this:
```
describe('some tests', () => {
  it('A - changes the user name', () => {
    const user = createUser()
    const result = changeUserName(user)
    expect(result.name).isEqualTo('NewUserName')
  }

  it('B - does not change the user name', () => {
    const user = createUser()
    const result = doSomethingThatWontChangeTheUserName(user)
    expect(result.name).isEqualTo('UserName')
  }

  function createUser(): User {
    return { name: 'UserName' }
  }
})
```
It's also possible to use the `beforeEach` to instantiate the User, but that do have some downside as well.

All of the above are some good stuff to be aware of you are writting tests in the future :)

File: frontend/src/admin/__tests__/user-list.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      }]
+    }
+  ]
+
+  it('renders user list correctly', () => {
+    const {getByText} = render(<UserList users={users} onClick={() => {
+    }}/>)
+    expect(getByText('Users')).toBeInTheDocument()
+    expect(getByText('John Doe')).toBeInTheDocument()
+    expect(getByText('Jane Doe')).toBeInTheDocument()
```
Comment (LindvedKrvang) [inline frontend/src/admin/__tests__/user-list.spec.tsx]: Minor thing: To allow you the flexibility of changing your test data without you having to change your tests you can write your assertions like this:
```
expect(getByText(users[0].name)).toBeInTheDocument()
expect(getByText(users[1].name)).toBeInTheDocument()
```
This also gets rid of two magic strings ;)

File: frontend/src/admin/admin-page.tsx [hunk 7]
```diff
@@ -1,13 +1,25 @@
...
+import './admin-page.scss'
 
 export default function AdminPage() {
-    return (
-        <div className="c-admin-page">
-            <UserList/>
-            <RestartServerButton/>
-        </div>
-    )
+  const [selectedUser, setSelectedUser] = useState<User | null>(null)
```
Comment (LindvedKrvang) [inline frontend/src/admin/admin-page.tsx]: Just write:
`const [selectedUser, setSelectedUser] = useState<User>()`
this makes the type `User | undefined` which is nicer to work with in Typescript.

Also, everytime you have to parse `undefined` or `null` to something you should take a deep inhale (preferably through your nose) so you can smell the odor of the giant code smell. And if that is not enough you should look behind you to see the giant red flag ;)

File: frontend/src/admin/components/access-rights/access-rights.tsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+export default function AccessRights({ selectedUser }: AccessRightsProps) {
+  const [accessRightGroups, setAccessRightGroups] = useState(selectedUser.accessRightGroups || [])
+
+  const handleCreateGroup = () => {
+    const newGroup = {
+      name: 'New Group',
+      accessRights: []
+    }
+    setAccessRightGroups([...accessRightGroups, newGroup])
+    console.log(newGroup)
```
Comment (LindvedKrvang) [inline frontend/src/admin/components/access-rights/access-rights.tsx]: Delete console logs :)

Comment (LindvedKrvang) [reply frontend/src/admin/components/access-rights/access-rights.tsx]: Also, this method appears to do nothing. I assume it's because it's a work in progress? :)

File: frontend/src/admin/components/dropdown-menu/dropdown-menu.tsx [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+  function hideDropdown(event: React.FocusEvent<HTMLButtonElement>): void {
+    if (event.currentTarget === event.target) {
+      setShowDropdown(false)
+    }
+  }
+
+  return (
+    <div className={showDropdown
+      ? 'c-dropdown-menu--active'
```
Comment (LindvedKrvang) [inline frontend/src/admin/components/dropdown-menu/dropdown-menu.tsx]: When the ternary is this small, it's fine to keep it on one line.
Also, consider if you want to write it like this: 
```
<div className={`c-dropdown-menu${showDropdown ? '--active' : ''}`}>
```
It's down to personal preference :)

File: frontend/src/admin/components/dropdown-menu/dropdown-menu.tsx [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+export default function DropdownMenu( {children}: DropdownMenuProps) {
+  const [showDropdown, setShowDropdown] = useState(false)
+
+  function toggleDropdown() {
+    setShowDropdown(!showDropdown)
+  }
+
+  function hideDropdown(event: React.FocusEvent<HTMLButtonElement>): void {
+    if (event.currentTarget === event.target) {
```
Comment (LindvedKrvang) [inline frontend/src/admin/components/dropdown-menu/dropdown-menu.tsx]: Consider using a form of defensive programming to avoid having the main part your function being nested:
```
function hideDropdown(...): void {
  if (evet.currentTarget !== event.target) {
    return
  }
  setShowDropdown(false)
}
```

File: frontend/src/admin/components/editable-table/editable-table.tsx [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+import {AccessRightGroup} from '../../models/access-right-group'
+import './editable-table.scss'
+
+export default function EditableTable({ accessRightGroups }: { accessRightGroups: AccessRightGroup[] }) {
+  const [editMode, setEditMode] = useState(false)
+
+  const [selectedRow, setSelectedRow] = useState<{
+    groupId: number;
+    accessRightId: number;
+  } | null>(null)
```
Comment (LindvedKrvang) [inline frontend/src/admin/components/editable-table/editable-table.tsx]: Refer to my other comment where you pass `null` to `useState`

File: frontend/src/admin/components/editable-table/editable-table.tsx [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+import EditRowIcon from '../edit-row-icon/edit-row-icon'
+import DeleteRowIcon from '../delete-row-icon/delete-row-icon'
+import SaveEditIcon from '../save-edit-icon/save-edit-icon'
+import CancelEditIcon from '../cancel-edit-icon/cancel-edit-icon'
+import {AccessRight} from '../../models/access-right'
+import {AccessRightGroup} from '../../models/access-right-group'
+import './editable-table.scss'
+
+export default function EditableTable({ accessRightGroups }: { accessRightGroups: AccessRightGroup[] }) {
+  const [editMode, setEditMode] = useState(false)
```
Comment (LindvedKrvang) [inline frontend/src/admin/components/editable-table/editable-table.tsx:11]: This technically acts as a boolean flag which is a code smell, since boolean flags indicate more than one responsibility.

So this component both have the responsibility of showing a table and being able to edit a table. This forces you have the `editMode` check through the component which is not nice.

You should consider having one component for just showing the table and then one for being able to edit it.

File: frontend/src/admin/components/editable-table/editable-table.tsx [hunk 13]
```diff
@@ -0,0 +1,189 @@
+import {useState} from 'react'
+import EditRowIcon from '../edit-row-icon/edit-row-icon'
+import DeleteRowIcon from '../delete-row-icon/delete-row-icon'
+import SaveEditIcon from '../save-edit-icon/save-edit-icon'
+import CancelEditIcon from '../cancel-edit-icon/cancel-edit-icon'
+import {AccessRight} from '../../models/access-right'
+import {AccessRightGroup} from '../../models/access-right-group'
+import './editable-table.scss'
+
+export default function EditableTable({ accessRightGroups }: { accessRightGroups: AccessRightGroup[] }) {
```
Comment (LindvedKrvang) [inline frontend/src/admin/components/editable-table/editable-table.tsx:10]: You have named your component `EditableTable` which doesn't indicate a specific use so the logical assumption is that this component can be used whenever you need a table.
However, the implementation of the component is heavily tied to AccessRightGroups, so `EditableTable` is not a fitting name.
Instead your could name it `AccessRights`, `AccessRightGroups`, `AccessRightsOverview`, or something fourth. 
Notice how the name suggestions don't refer to any implementation details (table etc...)? This way you can change the implementation of the component without needing change the name.

File: frontend/src/admin/components/editable-table/editable-table.tsx [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+    setSelectedRow(null)
+    setEditMode(false)
+  }
+
+  function saveAccessRightChange() {
+    setSelectedRow(null)
+    setEditMode(false)
+  }
+
+  return (
```
Comment (LindvedKrvang) [inline frontend/src/admin/components/editable-table/editable-table.tsx:63]: Lots of "if this then render x, if not render y" checks in this HTML. We should have a talk about writing less of that

File: frontend/src/admin/components/user-list/user-list.tsx [hunk 15]
```diff
@@ -1,28 +1,52 @@
...
-            </div>
-        </section>
-    )
+interface UserListProps {
+    users: User[],
+    onClick: (user: User) => (void),
+}
+
+export default function UserList({ users, onClick }: UserListProps) {
+  const [selectedUser, setSelectedUser] = useState<User | null>(null)
```
Comment (LindvedKrvang) [inline frontend/src/admin/components/user-list/user-list.tsx]: Refer to my comment about parsing null

**Review by LindvedKrvang — COMMENTED** [review]
Overall it looks good. I left some comments that needs to be addressed.

Regarding your tests. You don't have many tests that test your actually  functionality, but I guess that's because you don't have much functionality yet. And it's hard to test something that isn't there ;)

File: frontend/src/admin/__tests__/dropdown-menu.spec.tsx (hunk 1 above)
Comment (idax6797) [reply frontend/src/admin/__tests__/dropdown-menu.spec.tsx:24]: This is now implemented. After the second click it check that the --active modifier is not in the document ✅

File: frontend/src/admin/__tests__/user-information.spec.tsx (hunk 3 above)
Comment (idax6797) [reply frontend/src/admin/__tests__/user-information.spec.tsx:35]: Of course, this is implemented in all tests. ✅

File: frontend/src/admin/__tests__/user-information.spec.tsx (hunk 2 above)
Comment (idax6797) [reply frontend/src/admin/__tests__/user-information.spec.tsx:31]: Correct, I have implemented an expect(), that checks if the text 'User Information' is in the document. ✅

File: frontend/src/admin/__tests__/user-list-item.spec.tsx (hunk 4 above)
Comment (idax6797) [reply frontend/src/admin/__tests__/user-list-item.spec.tsx:23]: Correct, I have implemented an expect(), that checks if the button is in the document. ✅

File: frontend/src/admin/__tests__/user-list.spec.tsx (hunk 6 above)
Comment (idax6797) [reply frontend/src/admin/__tests__/user-list.spec.tsx]: Haha, great no magic strings here - this i implemented. ✅

File: frontend/src/admin/__tests__/user-information.spec.tsx (hunk 5 above)
Comment (idax6797) [reply frontend/src/admin/__tests__/user-information.spec.tsx]: Arh great explanation. I have tried to implement the createUser() as in the example. ✅

File: frontend/src/admin/admin-page.tsx (hunk 7 above)
Comment (idax6797) [reply frontend/src/admin/admin-page.tsx]: This is now changed and the smell/red flag metaphor is understood. 😆 ✅

File: frontend/src/admin/components/access-rights/access-rights.tsx (hunk 8 above)
Comment (idax6797) [reply frontend/src/admin/components/access-rights/access-rights.tsx]: Yes it is work in progress, but the console.log is deleted. ✅

File: frontend/src/admin/components/dropdown-menu/dropdown-menu.tsx (hunk 9 above)
Comment (idax6797) [reply frontend/src/admin/components/dropdown-menu/dropdown-menu.tsx]: This is Implemented. ✅

File: frontend/src/admin/components/dropdown-menu/dropdown-menu.tsx (hunk 10 above)
Comment (idax6797) [reply frontend/src/admin/components/dropdown-menu/dropdown-menu.tsx]: Yes, i have now implemented the 'event.currentTarget !== event.target' . ✅

File: frontend/src/admin/components/editable-table/editable-table.tsx (hunk 11 above)
Comment (idax6797) [reply frontend/src/admin/components/editable-table/editable-table.tsx]: No bad smell here, this is changed. ✅

File: frontend/src/admin/components/editable-table/editable-table.tsx (hunk 12 above)
Comment (idax6797) [reply frontend/src/admin/components/editable-table/editable-table.tsx:11]: I see. This is properly being changed in the next PR. Here the edit mode is being toggled via a button instead of toggling the edit mode on each row.

File: frontend/src/admin/components/user-list/user-list.tsx (hunk 15 above)
Comment (idax6797) [reply frontend/src/admin/components/user-list/user-list.tsx]: This is changed ✅

File: frontend/src/styles/app.scss [hunk 16]
```diff
@@ -1,14 +1,7 @@
...
-  grid-template-rows: repeat(5, 1fr);
-  grid-column-gap: 24px;
-  grid-row-gap: 24px;
-  padding: 40px;
-  background-color: $background-color;
-  color: $contrast-color;
-  font-family: Helvetica Neue, sans-serif;
-  font-size: 16px;
+  max-height: 100%;
+  margin: 0;
```
Comment (KvelaGorrrrnio) [inline frontend/src/styles/app.scss]: Is it necessary to set `margin: 0;` for the app?

File: frontend/src/admin/__tests__/user-list-item.spec.tsx [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+    }]}
+
+  it('renders without crashing', () => {
+    const onClick = jest.fn()
+    render(<UserListItem user={user} onClick={onClick} isSelected={false} />)
+    expect(screen.getByRole('button')).toBeInTheDocument()
+  })
+
+  it('should call onClick when the button is clicked', async () => {
+    const onClickMock = jest.fn() // create a mock function
```
Comment (KvelaGorrrrnio) [inline frontend/src/admin/__tests__/user-list-item.spec.tsx:28]: The comment can be removed :)

File: frontend/src/admin/__tests__/user-list.spec.tsx [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+    expect(getByText('Users')).toBeInTheDocument()
+    expect(getByText(users[0].name)).toBeInTheDocument()
+    expect(getByText(users[1].name)).toBeInTheDocument()
+  })
+
+  it('clicking "Download user" button calls downloadUser function', () => {
+    const downloadUserMock = window.alert = jest.fn()
+    const { getByText } = render(<UserList users={users} onClick={downloadUserMock} />)
+    fireEvent.click(getByText('Download user'))
+    expect(downloadUserMock).toHaveBeenCalled()
```
Comment (KvelaGorrrrnio) [inline frontend/src/admin/__tests__/user-list.spec.tsx:50]: It is good practice to be specific about what you expect. In this case: How many times to you expect the `downloadUserMock` to have been called? In this case it could potentially have been called a million times and the test will still pass.

File: frontend/src/admin/components/clone-user-access/clone-user-access.scss [hunk 19]
```diff
@@ -0,0 +1,6 @@
+@import "../../../styles/main";
+
+.c-clone-user-access {
+  max-width: 200px;
+  margin: 0 0 24px;
+}
```
Comment (KvelaGorrrrnio) [inline frontend/src/admin/components/clone-user-access/clone-user-access.scss:6]: This styling does not seem to be related to the `CloneUserAcces`, but rather the layout that it resides in.

File: frontend/src/admin/components/clone-user-access/clone-user-access.tsx [hunk 20]
```diff
@@ -0,0 +1,8 @@
+import './clone-user-access.scss'
+
+export default function CloneUserAccess() {
```
Comment (KvelaGorrrrnio) [inline frontend/src/admin/components/clone-user-access/clone-user-access.tsx:3]: Is it necessary to have this component?
And if it is needed: The component name is phrased as if it is an action, then perhaps it should be changed to something like `CloneUserAccessButton`.

File: frontend/src/admin/components/cancel-edit-icon/cancel-edit-icon.tsx [hunk 21]
```diff
@@ -0,0 +1,9 @@
+import React from 'react'
+
+export default function CancelEditIcon() {
```
Comment (KvelaGorrrrnio) [inline frontend/src/admin/components/cancel-edit-icon/cancel-edit-icon.tsx:3]: You are naming the component based on what it used for and not on what is.
If the icon used for the cancel action is a cross, then perhaps this should be called `CrossIcon`.
Then it can be used for other things than just exiting edit mode.

The same is true for the other icons.

File: frontend/src/admin/components/dropdown-menu/dropdown-menu.scss [hunk 22]
```diff
@@ -0,0 +1,46 @@
+@import "../../../styles/main";
+
+.c-dropdown-menu {
+  position: relative;
+
+  button {
```
Comment (KvelaGorrrrnio) [inline frontend/src/admin/components/dropdown-menu/dropdown-menu.scss]: This targets all buttons below in the dropdown menu. Is that intended behaviour or is it a specific button/ set of buttons that should be targeted?

File: frontend/src/admin/components/dropdown-menu/dropdown-menu.scss [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+  &--active {
+    position: relative;
+
+    button {
+      border: none;
+      margin: 0;
+      padding: 0;
+    }
+
+    .c-dropdown-menu-items {
```
Comment (KvelaGorrrrnio) [inline frontend/src/admin/components/dropdown-menu/dropdown-menu.scss]: This is a separate component than the `.c-dropdown-menu`, and should have its own file. By doing this you also encourage the independence of the two components.

File: frontend/src/admin/components/editable-table/editable-table.scss [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+          &:nth-of-type(2):before {
+            content: "Label:";
+          }
+
+          &:nth-of-type(3):before {
+            content: "Anonymous:";
+          }
+          &:nth-of-type(4):before {
+            content: "Path and Args:";
+          }
```
Comment (KvelaGorrrrnio) [inline frontend/src/admin/components/editable-table/editable-table.scss:139]: This should be modeled in JSX. If we are to make a translation feature, then the translation of this will be quite cumbersome.

File: frontend/src/admin/components/editable-table/editable-table.scss [hunk 25]
```diff
@@ -0,0 +1,181 @@
...
+
+          &:before {
+            margin-right: 1rem;
+          }
+
+          &:last-child {
+            justify-content: flex-start !important;
```
Comment (KvelaGorrrrnio) [inline frontend/src/admin/components/editable-table/editable-table.scss:146]: Why is `!important` used here?

File: frontend/package.json [hunk 26]
```diff
@@ -9,23 +9,36 @@
...
     "lint-fix": "eslint . --ext .ts,.tsx --fix",
-    "test": "echo \"no test setup is configured.\"",
-    "clean": "rimraf dist"
+    "clean": "rimraf dist",
+    "test": "jest"
   },
   "devDependencies": {
+    "@babel/core": "^7.21.3",
+    "@babel/preset-env": "^7.20.2",
+    "@babel/preset-typescript": "^7.21.0",
```
Comment (KvelaGorrrrnio) [inline frontend/package.json:18]: Is babel necessary?

File: frontend/src/admin/components/access-rights/access-rights.tsx [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+import {User} from '../../models/user'
+
+interface AccessRightsProps {
+  selectedUser: User;
+}
+
+export default function AccessRights({ selectedUser }: AccessRightsProps) {
+  const [accessRightGroups, setAccessRightGroups] = useState(selectedUser.accessRightGroups || [])
+
+  const handleCreateGroup = () => {
```
Comment (KvelaGorrrrnio) [inline frontend/src/admin/components/access-rights/access-rights.tsx:13]: `handle` can be remove. You could also specify the state of the group that is created. Perhaps `createEmptyAccessRightGroup`?

**Review by LindvedKrvang — APPROVED** [review]

**Review by KvelaGorrrrnio — APPROVED** [review]
There are still things that needs to be addressed. But that will be at a later stage.
